File: app/src/main/java/com/adasa/foodplan/data/local/dao/MealPlanDao.kt
package com.adasa.foodplan.data.local.dao

import androidx.room.*
import com.adasa.foodplan.data.local.entity.DayPlanEntity
import com.adasa.foodplan.data.local.entity.MealPlanEntity
import com.adasa.foodplan.data.local.entity.MealSlotEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface MealPlanDao {
    @Query("SELECT * FROM meal_plans ORDER BY startDate ASC")
    fun getAllMealPlans(): Flow<List<MealPlanEntity>>

    @Query("SELECT * FROM meal_plans WHERE id = :id")
    suspend fun getMealPlanById(id: String): MealPlanEntity?

    @Query("SELECT * FROM day_plans WHERE mealPlanId = :mealPlanId ORDER BY date ASC")
    suspend fun getDayPlansForMealPlan(mealPlanId: String): List<DayPlanEntity>

    @Query("SELECT * FROM meal_slots WHERE dayPlanId = :dayPlanId")
    suspend fun getMealSlotsForDayPlan(dayPlanId: String): List<MealSlotEntity>

    @Upsert
    suspend fun upsertMealPlan(mealPlan: MealPlanEntity)

    @Upsert
    suspend fun upsertDayPlan(dayPlan: DayPlanEntity)

    @Upsert
    suspend fun upsertMealSlot(mealSlot: MealSlotEntity)

    @Delete
    suspend fun deleteMealPlan(mealPlan: MealPlanEntity)

    // Epoch-day params avoid TypeConverter issues with query parameters
    @Query("SELECT * FROM day_plans WHERE date = :dateEpoch LIMIT 1")
    suspend fun getDayPlanByDate(dateEpoch: Long): DayPlanEntity?

    @Query("SELECT * FROM day_plans WHERE date >= :startEpoch AND date <= :endEpoch ORDER BY date ASC")
    suspend fun getDayPlansInRange(startEpoch: Long, endEpoch: Long): List<DayPlanEntity>

    @Query("SELECT * FROM meal_slots WHERE dayPlanId IN (:dayPlanIds)")
    suspend fun getMealSlotsForDayPlanIds(dayPlanIds: List<String>): List<MealSlotEntity>
}

File: flutter-app/lib/data/local/dao/meal_plan_dao.dart
import 'package:drift/drift.dart';
import 'package:foodplan/data/local/app_database.dart';

part 'meal_plan_dao.g.dart';

@DriftAccessor(
  tables: [MealPlansTable, DayPlansTable, MealSlotsTable],
)
class MealPlanDao extends DatabaseAccessor<AppDatabase>
    with _$MealPlanDaoMixin {
  MealPlanDao(super.db);

  // ── MealPlan ──────────────────────────────────────────────────────────────

  Stream<List<MealPlansTableData>> watchAllMealPlans() =>
      (select(mealPlansTable)
            ..orderBy([(t) => OrderingTerm.asc(t.startDate)]))
          .watch();

  Future<MealPlansTableData?> getMealPlanById(String id) =>
      (select(mealPlansTable)..where((t) => t.id.equals(id)))
          .getSingleOrNull();

  Future<void> upsertMealPlan(MealPlansTableCompanion row) =>
      into(mealPlansTable).insertOnConflictUpdate(row);

  Future<void> deleteMealPlan(String id) =>
      (delete(mealPlansTable)..where((t) => t.id.equals(id))).go();

  // ── DayPlan ───────────────────────────────────────────────────────────────

  Future<List<DayPlansTableData>> getDayPlansForMealPlan(
          String mealPlanId) =>
      (select(dayPlansTable)
            ..where((t) => t.mealPlanId.equals(mealPlanId))
            ..orderBy([(t) => OrderingTerm.asc(t.date)]))
          .get();

  Future<DayPlansTableData?> getDayPlanByDate(int epochDays) =>
      (select(dayPlansTable)..where((t) => t.date.equals(epochDays)))
          .getSingleOrNull();

  Future<void> upsertDayPlan(DayPlansTableCompanion row) =>
      into(dayPlansTable).insertOnConflictUpdate(row);

  /// Fetch day plans within [startEpochDays]..[endEpochDays] inclusive.
  Future<List<DayPlansTableData>> getDayPlansInRange(
          int startEpochDays, int endEpochDays) =>
      (select(dayPlansTable)
            ..where((t) =>
                t.date.isBiggerOrEqualValue(startEpochDays) &
                t.date.isSmallerOrEqualValue(endEpochDays))
            ..orderBy([(t) => OrderingTerm.asc(t.date)]))
          .get();

  // ── MealSlot ──────────────────────────────────────────────────────────────

  Future<List<MealSlotsTableData>> getMealSlotsForDayPlan(
          String dayPlanId) =>
      (select(mealSlotsTable)
            ..where((t) => t.dayPlanId.equals(dayPlanId))
            ..orderBy([(t) => OrderingTerm.asc(t.slotIndex)]))
          .get();

  Future<void> upsertMealSlot(MealSlotsTableCompanion row) =>
      into(mealSlotsTable).insertOnConflictUpdate(row);

  Future<void> deleteMealSlotsForDayPlan(String dayPlanId) =>
      (delete(mealSlotsTable)
            ..where((t) => t.dayPlanId.equals(dayPlanId)))
          .go();

  Future<List<MealSlotsTableData>> getMealSlotsForDayPlanIds(
          List<String> dayPlanIds) =>
      (select(mealSlotsTable)
            ..where((t) => t.dayPlanId.isIn(dayPlanIds))
            ..orderBy([(t) => OrderingTerm.asc(t.slotIndex)]))
          .get();

  /// Fetch meal plans whose dates fall within [startEpochDays]..[endEpochDays].
  Future<List<MealPlansTableData>> getMealPlansInRange(
          int startEpochDays, int endEpochDays) =>
      (select(mealPlansTable)
            ..where((t) =>
                t.startDate.isBiggerOrEqualValue(startEpochDays) &
                t.endDate.isSmallerOrEqualValue(endEpochDays)))
          .get();
}
